test/e2e/aks_webhooks.go

```diff
@@ -0,0 +1,391 @@
+		}
+		amcp.Spec.APIServerAccessProfile = apiServerAccessProfile
+		err = mgmtClient.Update(ctx, amcp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+	}
+
+	By("Validating that AzureManagedMachinePool webhooks are working")
+	var nodePoolModeToUserError error
+	for _, mp := range input.MachinePools {
+		ammp, err := getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		By(fmt.Sprintf("Validating webhooks against AzureManagedMachinePool %s", ammp.Name))
+
+		By("Validating that we can't add a node label using the reserved 'kubernetes.azure.com/' prefix")
+		if ammp.Spec.NodeLabels == nil {
+			ammp.Spec.NodeLabels = map[string]string{}
+		}
+		ammp.Spec.NodeLabels[azureutil.AzureSystemNodeLabelPrefix+"/foo"] = "bar"
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.OSType")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if to.String(ammp.Spec.OSType) == "Linux" {
+			ammp.Spec.OSType = to.StringPtr("Windows")
+		} else {
+			ammp.Spec.OSType = to.StringPtr("Linux")
+		}
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.SKU")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		ammp.Spec.SKU = ammp.Spec.SKU + "_vInvalid"
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't update AzureManagedMachinePool.Spec.OSDiskSizeGB")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		ammp.Spec.OSDiskSizeGB = to.Int32Ptr(to.Int32(ammp.Spec.OSDiskSizeGB) + 1)
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't add a new 'infrastructure.cluster.x-k8s.io/custom-header-' Annotation")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if ammp.ObjectMeta.Annotations == nil {
+			ammp.ObjectMeta.Annotations = map[string]string{}
+		}
+		ammp.ObjectMeta.Annotations[azure.CustomHeaderPrefix+"foo"] = "bar"
+		err = mgmtClient.Update(ctx, ammp)
+		Expect(err).To(HaveOccurred())
+		Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+		By("Validating that we can't add or remove from AzureManagedMachinePool.Spec.AvailabilityZones")
+		ammp, err = getAzureManagedMachinePool(ctx, mgmtClient, mp)
+		Expect(err).NotTo(HaveOccurred())
+		if ammp.Spec.AvailabilityZones == nil {
+			ammp.Spec.AvailabilityZones = []string{}
+		}
+		ammp.Spec.AvailabilityZones = append(ammp.Spec.AvailabilityZones, "99")
```

Comment:
FWIW I think this nil check can be omitted since appending to a nil slice directly works the same way: https://go.dev/play/p/m5rP1TWFiDU

Reply:
lol I had PTSD from the `map[string]string` above